### PR TITLE
Fix some more cases in ConformanceCollector.

### PR DESCRIPTION
**Pull request by eeckstein**
Most important: consider resilient types (IRGen needs the metadata for a resilient type or a type which contains a resilient type).

No test cases added yet. It's hard to find isolated test cases, but all things are tested somewhere on the bots.
I'll try to add some isolated tests later.

rdar://problem/30250350

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci Please test and merge

File: lib/SIL/InstructionUtils.cpp [hunk 1]
```diff
@@ -246,9 +246,52 @@ void ConformanceCollector::scanType(Type type) {
...
+    if (SubType->hasArchetype())
+      return true;
+
+    if (NominalTypeDecl *NT = SubType->getNominalOrBoundGenericNominal()) {
+      if (!NT->hasFixedLayout(M.getSwiftModule(), ResilienceExpansion::Maximal))
+        return true;
+
+      // Check the types of all stored properties of the nominal type.
+      for (VarDecl *Field : NT->getStoredProperties()) {
+        if (Field->hasClangNode() || !Field->hasInterfaceType())
```
Comment (slavapestov) [inline lib/SIL/InstructionUtils.cpp]: Have you seen fields without an interface type here? That should not happen after type checking AFAIK.

Also if a field has a clang node, it means the overall type has a clang node, so you can handle that outside the loop.

File: lib/SIL/InstructionUtils.cpp [hunk 2]
```diff
@@ -246,9 +246,52 @@ void ConformanceCollector::scanType(Type type) {
...
+          continue;
+        CanType FieldTy = SubType->getTypeOfMember(M.getSwiftModule(),
```
Comment (slavapestov) [inline lib/SIL/InstructionUtils.cpp]: If you make this take a SILType, you can use SILType::getFieldType(). It's a bit simpler.

File: lib/SIL/InstructionUtils.cpp [hunk 3]
```diff
@@ -401,12 +444,22 @@ void ConformanceCollector::collect(swift::SILInstruction *I) {
     case ValueKind::SelectEnumAddrInst:
     case ValueKind::IndexAddrInst:
     case ValueKind::RefElementAddrInst:
-    case ValueKind::CopyAddrInst: {
+    case ValueKind::CopyAddrInst:{
       Type Ty = I->getOperand(0)->getType().getSwiftRValueType();
```
Comment (slavapestov) [inline lib/SIL/InstructionUtils.cpp]: Pass a SILType here instead of unwrapping it with getSwiftRValueType()

File: lib/SIL/InstructionUtils.cpp [hunk 4]
```diff
@@ -401,12 +444,22 @@ void ConformanceCollector::collect(swift::SILInstruction *I) {
...
-      if (Ty->hasArchetype())
+      if (needsMetaData(Ty))
         scanType(Ty);
       break;
     }
+    case ValueKind::AllocGlobalInst: {
+      SILGlobalVariable *var = cast<AllocGlobalInst>(I)->getReferencedGlobal();
+      scanType(var->getLoweredType().getSwiftRValueType());
```
Comment (slavapestov) [inline lib/SIL/InstructionUtils.cpp]: And here

File: lib/SIL/InstructionUtils.cpp [hunk 5]
```diff
@@ -401,12 +444,22 @@ void ConformanceCollector::collect(swift::SILInstruction *I) {
...
+      break;
+    }
+    case ValueKind::GlobalAddrInst: {
+      SILGlobalVariable *var = cast<GlobalAddrInst>(I)->getReferencedGlobal();
+      scanType(var->getLoweredType().getSwiftRValueType());
```
Comment (slavapestov) [inline lib/SIL/InstructionUtils.cpp]: And here

File: lib/SIL/InstructionUtils.cpp [hunk 6]
```diff
@@ -246,9 +246,52 @@ void ConformanceCollector::scanType(Type type) {
...
+          if (needsMetaData(FieldTy))
+            scanType(FieldTy);
+        }
+      }
+    }
+  });
+}
+
+bool ConformanceCollector::needsMetaData(Type type) {
+  type = type->getCanonicalType();
```
Comment (slavapestov) [inline lib/SIL/InstructionUtils.cpp]: Change this to SILType to remove unnecessary getCanonicalType() calls here and elsewhere

File: lib/SIL/InstructionUtils.cpp [hunk 7]
```diff
@@ -246,9 +246,52 @@ void ConformanceCollector::scanType(Type type) {
         auto substs = SubType->gatherAllSubstitutions(M.getSwiftModule(),
                                                      nullptr);
         scanSubsts(substs);
+
+        // Check the types of all stored properties of the nominal type.
```
Comment (slavapestov) [inline lib/SIL/InstructionUtils.cpp]: Can you replace this loop with a needsMetaData() call?

File: include/swift/SIL/InstructionUtils.h [hunk 8]
```diff
@@ -94,9 +94,18 @@ class ConformanceCollector {
   /// Conformances and types which have been handled.
   llvm::SmallPtrSet<const void *, 8> Visited;
 
+  llvm::DenseMap<TypeBase *, bool> TypeNeedsMetaData;
```
Comment (slavapestov) [inline include/swift/SIL/InstructionUtils.h]: Nitpick: Elsewhere we spell this as "Metadata", not capitalizing the D

File: lib/SIL/InstructionUtils.cpp (hunk 1 above)
Comment (eeckstein) [inline lib/SIL/InstructionUtils.cpp]: Yes, the IRGen/global_resilience.sil comes up with a missing interface type (when compiling with  --swift-stdlib-enable-resilience)

Comment (eeckstein) [inline lib/SIL/InstructionUtils.cpp]: If you want to investigate this, remove the isInterfaceType check here and it will crash in IRGen

**Comment by eeckstein** [comment]
@slavapestov Thanks for reviewing!

File: lib/SIL/InstructionUtils.cpp (hunk 1 above)
Comment (slavapestov) [inline lib/SIL/InstructionUtils.cpp]: Can you dump the property and see? This isn't supposed to happen, we might be missing something if the property is not type checked.

File: lib/SIL/InstructionUtils.cpp (hunk 7 above)
Comment (eeckstein) [inline lib/SIL/InstructionUtils.cpp]: No, because I have to record all conformances and types

File: lib/SIL/InstructionUtils.cpp (hunk 1 above)
Comment (eeckstein) [inline lib/SIL/InstructionUtils.cpp]: (lldb) p Field->dump()
(var_decl "w" type='<null type>' storage_kind=stored)
(lldb) p SubType->dump()
(struct_type decl=resilient_struct.(file).Size@/Users/erik/proj/swift-work/swift/test/IRGen/../Inputs/resilient_struct.swift:16:15)

Comment (eeckstein) [inline lib/SIL/InstructionUtils.cpp]: `(lldb) p Field->dump()
(var_decl "w" type='<null type>' storage_kind=stored)
(lldb) p SubType->dump()
(struct_type decl=resilient_struct.(file).Size@/Users/erik/proj/swift-work/swift/test/IRGen/../Inputs/resilient_struct.swift:16:15)
`

Comment (eeckstein) [inline lib/SIL/InstructionUtils.cpp]: PS: I removed the angle brackets around the "null type"

File: lib/SIL/InstructionUtils.cpp (hunk 6 above)
Comment (eeckstein) [inline lib/SIL/InstructionUtils.cpp]: Well, I tried this. But it adds more complications than simplifications. But what I did is that I changed the parameter type from Type to CanType.

**Comment by eeckstein** [comment]
@swift-ci Please test

**Comment by eeckstein** [comment]
@swift-ci Please test

**Comment by eeckstein** [comment]
@swift-ci Please test

**Pull request closed without merging**